src/kaleidoscope/EventQueue.h

```diff
@@ -0,0 +1,105 @@
+// -*- mode: c++ -*-
+/* Kaleidoscope - Firmware for computer input devices
+ * Copyright (C) 2013-2019  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <Arduino.h>
+
+#include "kaleidoscope/Kaleidoscope.h"
+#include "kaleidoscope/KeyAddr.h"
+
+namespace kaleidoscope {
+
+// This class defines a generic event queue that stores both key press and
+// release events, recording the key address, a timestamp, and the keyswitch
+// state (press or release).
+template <uint8_t _max_length,
+          typename _Bitfield  = uint8_t,
+          typename _Timestamp = uint16_t>
+class EventQueue {
+
+  static_assert(_max_length <= 255,
+                "EventQueue error: _max_length must be less than 256!");
+  static_assert(_max_length <= (sizeof(_Bitfield) * 8),
+                "EventQueue error: _Bitfield type too small for _max_length!");
+
+ private:
+  uint8_t    length_{0};
+  KeyAddr    addrs_[_max_length];
+  _Timestamp timestamps_[_max_length];
+  _Bitfield  release_event_bits_;
+
+ public:
+  uint8_t length() const { return length_; }
+  bool    isEmpty() const { return (length_ == 0); }
+  bool    isFull() const { return (length_ == _max_length); }
+
+  KeyAddr addr(uint8_t index) const { return addrs_[index]; }
+
+  _Timestamp timestamp(uint8_t index) const { return timestamps_[index]; }
+
+  bool isRelease(uint8_t index) const {
+    return bitRead(release_event_bits_, index);
+  }
+  bool isPress(uint8_t index) const { return !isRelease(index); }
+
+  // Append a new event on the end of the queue.
+  void append(KeyAddr k, uint8_t keyswitch_state) {
+    addrs_[length_]      = k;
+    timestamps_[length_] = Kaleidoscope.millisAtCycleStart();
+    bitWrite(release_event_bits_, length_, keyToggledOff(keyswitch_state));
+    ++length_;
+  }
+
+  // Remove the first event from the head of the queue, shifting the others.
+  void shift() {
+    --length_;
+    for (uint8_t i{0}; i < length_; ++i) {
```

Comment:
Could shift and remove be possibly made more performant by using a [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer) approach instead? This would remove the copying operations.

Reply:
I would guess so, but the downside is that `EventQueue` would then need to store two pointers or indices, and do more computation every time it iterates through the queue, which Qukeys does much more often than it shifts the queue. I did consider trying to use a circular buffer, but this seemed to be more likely the better choice overall. I'm not at all opposed to using a circular buffer if it turns out to be better; I just did it this way because the overall code was more obvious to me.

Reply:
A circular buffer variant would need only one additional modulo operation per loop cycle when iterating events and, as you pointed out, a single additional byte for the second index.

Reply:
I would like to see how that is accomplished.

Reply:
Wouldn't the ring buffer also need to translate the index value whenever one of the stored items is requested?

Reply:
Assuming that `EventQueue` would have methods `begin()`, `end()` and `capacity()`
```cpp
for(auto i = event_queue_.begin(); i != event_queue_.end(); ++i, i %= event_queue_.capacity()) {
   ...
}
```
`end()` would point to the element that follows the actual last element.

Could could also implement an iterator that wraps this up to do range based for looping.

Reply:
FWIW, this is especially performant if capacity is a power of two (like in your case with capacity == 8) because it enables the compiler to turn the modulo into a simple and operation with a bitmask.

Reply:
I was not considering exposing the internal values to the caller like that. I was also under the impression that modulo operations were somewhat expensive on AVR processors. I guess I should figure out how to use the compiler explorer…

The range-based iterator seems like a good idea, but that makes it harder to start searching anywhere but the head of the queue, which is another thing that Qukeys does in some cases.

Reply:
Most loops in `Qukeys.cpp` are based on `length()` which already somewhat exposes 'internal information'.

Regarding range-based iterators:
I do not see any reason why the caller side shouldn't be able to choose between random (index based) access and range based iteration. It simply depends on the application which one is the more suitable choice. STL's std::vector, e.g., provides both for good reason.

I really don't want to push you with respect to circular buffers but just wanted to point out that you could save quite a bit of PROGMEM by eliminating great parts of `shift()` and `remove()` for the cost of one byte of RAM and approx. four additional assembly instructions for the modulo stuff.

Reply:
I will definitely give it a try as soon as I get the time to write it up, but the cost that concerns me more is the added complexity in the calling _code_. I will also be surprised if the PROGMEM savings are very substantial. Qukeys doesn't actually use `remove()`, which would require the same amount of copying, even with the ring buffer.

Reply:
I gave it a try, using an implementation that eposes the internal `head()` and `tail()` index values to the caller. It came in at +1 byte of RAM, -4 bytes PROGMEM, neither of which is a particularly big difference. Take a look at the [EventQueue-ring-buffer](https://github.com/gedankenexperimenter/Kaleidoscope/tree/EventQueue-ring-buffer) branch if you want to see it. [Note: at first, I thought it saved 8 bytes of PROGMEM, but that was before I noticed an error I made in the implementation.]

If this was intended for a very long queue, I would definitely want to use a ring buffer, and would be motivated to put more effort into making it nicer. It is unquestionably more efficient when shifting the queue, but in this application, the queue is shifted once per key event, but searched (when a qukey is in an indeterminate state) at least once per cycle, which is far, far more frequent. In addition, I find the resulting loop much more difficult to follow. Compare:
```c++
// ring buffer version
for (uint8_t i = event_queue_.head() + 1; i != event_queue_.tail(); ++i, i %= queue_capacity_) {
  // body
}
// length-based version
for (uint8_t i{1}; i < event_queue_.length(); ++i) {
  // body
}
```
The ring buffer version is much easier to get wrong. There are, of course, any number of things I could do to improve this. I tried adding code to keep the calling code the same, but that ended up adding more than 200 bytes of PROGMEM.

I really do like the idea of using a ring buffer for this, but I just don't find it practical compared to actually shifting the queue in this case. Keeping the head index at zero basically confines all of the messy code to the body of one function, rather than keeping that one function simple at the expense of nearly everything else involved. If that one function got called much more frequently, with a much longer expected queue size, it would clearly be a better choice to use the ring buffer, but I'm not seeing that happening here.

Reply:
@noseglasses – Given that Qukeys is not using `remove(index)`, do you think it would be best to just remove it entirely?

I wrote it because at one point I thought it would help resolve a particular corner case, but then I found a better way. I'm ambivalent about having it there – I don't have a concrete use case for it, but I know that it's not so trivial to get right because of the bitfield, so maybe it's good to have it there. On the other hand, if this tool is sitting there, available, someone might be tempted to use it unnecessarily (and re-ordering events because it's convenient, rather than necessary, is likely to cause tricky bugs).

Reply:
Thanks for giving it a try. In your use case where the iteration over the container is the more imporant aspect, it makes sense to have a special (optimized) implementation that copies when shifting.

In general you could go with the following to make the calling code more readable

```cpp
// ring buffer version
for (uint8_t i = event_queue_.head() + 1; i != event_queue_.tail(); i = event_queue_.next(i)) {
  // body
}
```
with
```cpp
uint8_t EventQueue::next(uint8_t i) {
   ++i;
   return i % capacity_;
}
```

About the `remove()` method: I would remove it from the interface at it is not part of a typical queue implementation to remove stuff from the middle, at least not if considering what FIFO means.

BTW, in other common libraries I've seen the terms `pushBack(element)` and `popFront()` for the typical queue operations, which I find pretty intuitive names.

Reply:
One more thing. I would prefer the name `KeyEventQueue` as the term 'event' allone is to unspecific, especially given the fact that algernon started working on supporting rotation-knob-things (forgot the name), so there might be other types of 'events' in the future.

Reply:
Excellent points, all; thank you. I'll rename the class to make it more specific, but I think `KeyEventQueue` is also not quite right, because this is a queue for before the `Key` value has been determined. I think that either `KeyswitchEventQueue` or `KeyAddrEventQueue` would be more apt.

In Kaleidoglyph, I found it useful to have two separate hooks: `onKeyswitchEvent()` and `onKeyEvent()`, with different treatment and guarantees. The `KeyswitchEvent` variant only gets events from physical keyswitches, and only gets each event once, whereas the `KeyEvent` hooks run in a loop that gets restarted whenever the `Key` value changes, and can get simulated events generated by plugins.

For that matter, I also think `handleKeyswitchEvent()` is a misnomer in Kaleidoscope. Its parameters don't necessarily represent physical key _switches_, and also represent _states_, not _events_ — a very important distinction when trying to debug Kaleidoscope code.

At one point, I recall you commenting on the overuse of the term "key" for too many different things, and you suggested something more specific for the `Key` class, but I can't recall what it was. I do remember being in favor of it, though.

Reply:
Regarding `pushBack()` & `popFront()`: I do not find those names particularly intuitive; though their meaning is clear, it takes me a few seconds each time I look at them to deduce their function. More important, they are not quite analogous to `append()` & (especially) `shift()`. For (arguably misguided) performance reasons, I haven't made a `QueueEntry` class, so there isn't a single object that is passed to `append()`, and there's no return type at all for `shift()`. The need for random access to the queue entries, and the fact that Qukeys frequently looks only at one property of each queue entry are what led me to implement it in this way. I feel like adding a `popFront()` method would require either an additional type to be defined as the return type. That's not too hard to do, but I still wouldn't use it for Qukeys, because by the time it knows that `shift()` should be called, it already has a copy of the event type and address, so there's no need to call `popFront()`.

This all suggests to me that `EventQueue` is more particular to Qukeys than I had first imagined it to be. I'll try adding an inner class to give `popFront()` & pushBack() an appropriate type, and see what it looks like, but I'm leaning towards leaving them out and putting this class header under `plugin/Qukeys/` to make it clear what it was intended for.

Reply:
> For that matter, I also think handleKeyswitchEvent() is a misnomer in Kaleidoscope. Its parameters don't necessarily represent physical key switches, and also represent states, not events — a very important distinction when trying to debug Kaleidoscope code.

Good point. Now as we have a flexible hook system, that due to recent changes supports deprecation of hooks, it would be possible to fix this without causing too much overhead.

> At one point, I recall you commenting on the overuse of the term "key" for too many different things, and you suggested something more specific for the Key class, but I can't recall what it was. I do remember being in favor of it, though.

Right, I remember that, too. IMHO, `KeyswitchEventQueue` would be a better choice than `KeyAddrEventQueue`, as we also use `KeyAddr` to reference per-key LED's, so the prefix `KeyAddr` does tell less about what the queue actually handles/stores.

> Regarding pushBack() & popFront(): ...

Letting the event queue store some sort of struct that represents the keyswitch event would make it more like a traditional container and `pushBack()` would make sense. But it is obvious that this would make using an internal bitfield more complex and possibly more costly.

FWIW, `popFront()` in some implementations/libraries does not return a value (cf. STL's deque or list). i have seen suggestions to rename it `dropFront()`, which would make more sense, though.

Reply:
> Letting the event queue store some sort of struct that represents the keyswitch event would make it more like a traditional container and `pushBack()` would make sense. But it is obvious that this would make using an internal bitfield more complex and possibly more costly.

Yeah, I'm not thrilled about the complexity of the bitfield implementation, but the savings of almost one byte per entry was enough to tempt me. On the other hand, that's the biggest thing that keeps the class from being more generally useful; on anything that wasn't so space constrained (e.g. ARM embedded processors), I think it would be better to just use a `bool` inside a structure used to store the event info. I'd still be inclined to keep the timestamp management separate, especially if I could figure out a way to have a template class specialization that doesn't use timestamps at all. And now that I've got that in my head, I'll give it a try.

In a similar vein (on account of the bitfield's specificity), I'm sold on preferring `KeyswitchEventQueue` to `KeyAddrEventQueue`. I had thought of the latter as being more illustrative of the data it stores (in light of the functions (mis)named `*KeyswitchEvent`), but it was the weakest of preferences. I'll change it; thanks.